File: modules/api/pkg/handler/v2/provider/baremetal.go
/*
Copyright 2024 The Kubermatic Kubernetes Platform contributors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package provider

import (
	"context"
	"fmt"
	"net/http"

	"github.com/go-kit/kit/endpoint"
	"github.com/gorilla/mux"
	providercommon "k8c.io/dashboard/v2/pkg/handler/common/provider"

	"k8c.io/dashboard/v2/pkg/provider"
	utilerrors "k8c.io/kubermatic/v2/pkg/util/errors"
)

// TinkerbellListImagesReq represents a request to list Tinkerbell images
// swagger:parameters listTinkerbellImages
type TinkerbellListImagesReq struct {
	// in: path
	// required: true
	DC string `json:"dc"`
}

func TinkerbellImagesEndpoint(userInfoGetter provider.UserInfoGetter, seedsGetter provider.SeedsGetter) endpoint.Endpoint {
	return func(ctx context.Context, request interface{}) (interface{}, error) {
		req, ok := request.(TinkerbellListImagesReq)
		if !ok {
			return nil, utilerrors.NewBadRequest("invalid request")
		}
		return providercommon.TinkerbellImages(ctx, req.DC, userInfoGetter, seedsGetter)
	}
}

func DecodeTinkerbellListImageReq(c context.Context, r *http.Request) (interface{}, error) {
	var req TinkerbellListImagesReq

	dc, ok := mux.Vars(r)["dc"]
	if !ok {
		return req, fmt.Errorf("'dc' parameter is required")
	}
	req.DC = dc

	return req, nil
}

File: modules/web/src/app/core/services/node-data/provider/baremetal.ts
// Copyright 2024 The Kubermatic Kubernetes Platform contributors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {NodeDataMode} from '@app/node-data/config';
import {ClusterSpecService} from '@core/services/cluster-spec';
import {NodeDataService} from '@core/services/node-data/service';
import {ProjectService} from '@core/services/project';
import {BaremetalService} from '@core/services/provider/baremetal';
import {TinkerbellOSImageList} from '@shared/entity/provider/baremetal';
import {NodeProvider} from '@shared/model/NodeProviderConstants';
import {Observable, of, onErrorResumeNext} from 'rxjs';
import {catchError, debounceTime, filter, switchMap, take, tap} from 'rxjs/operators';

export class NodeDataBaremetalProvider {
  private readonly _debounce = 500;

  constructor(
    private readonly _nodeDataService: NodeDataService,
    private readonly _baremetalService: BaremetalService,
    private readonly _clusterSpecService: ClusterSpecService,
    private readonly _projectService: ProjectService
  ) {}

  osImages(onError: () => void = undefined, onLoadingCb: () => void = null): Observable<TinkerbellOSImageList> {
    switch (this._nodeDataService.mode) {
      case NodeDataMode.Wizard:
        return this._clusterSpecService.datacenterChanges
          .pipe(debounceTime(this._debounce))
          .pipe(filter(_ => this._clusterSpecService.provider === NodeProvider.BAREMETAL))
          .pipe(tap(_ => (onLoadingCb ? onLoadingCb() : null)))
          .pipe(switchMap(datacenter => this._baremetalService.getOSImages(datacenter)))
          .pipe(
            catchError(_ => {
              if (onError) {
                onError();
              }

              return onErrorResumeNext(of({} as TinkerbellOSImageList));
            })
          );
      case NodeDataMode.Dialog: {
        return this._projectService.selectedProject
          .pipe(debounceTime(this._debounce))
          .pipe(tap(_ => (onLoadingCb ? onLoadingCb() : null)))
          .pipe(switchMap(_ => this._baremetalService.getOSImages(this._clusterSpecService.datacenter)))
          .pipe(
            catchError(_ => {
              if (onError) {
                onError();
              }

              return onErrorResumeNext(of({} as TinkerbellOSImageList));
            })
          )
          .pipe(take(1));
      }
    }
  }
}
